dot/state/grandpa_changes.go

```diff
@@ -0,0 +1,313 @@
+// this function updates the current state of the change tree
+func (oc *orderedPendingChanges) pruneChanges(hash common.Hash, isDescendantOf isDescendantOfFunc) error {
+	onBranchForcedChanges := make([]*pendingChange, 0, oc.Len())
+
+	for _, forcedChange := range *oc {
+		isDescendant, err := isDescendantOf(hash, forcedChange.announcingHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchForcedChanges = append(onBranchForcedChanges, forcedChange)
+		}
+	}
+
+	*oc = onBranchForcedChanges
+	return nil
+}
+
+type pendingChangeNode struct {
+	change *pendingChange
+	nodes  []*pendingChangeNode
+}
+
+func (c *pendingChangeNode) importNode(blockHash common.Hash, blockNumber uint, pendingChange *pendingChange,
+	isDescendantOf isDescendantOfFunc) (imported bool, err error) {
+	announcingHash := c.change.announcingHeader.Hash()
+
+	if blockHash.Equal(announcingHash) {
+		return false, fmt.Errorf("%w: %s", errDuplicateHashes, blockHash)
+	}
+
+	if blockNumber <= c.change.announcingHeader.Number {
+		return false, nil
+	}
+
+	for _, childrenNodes := range c.nodes {
+		imported, err := childrenNodes.importNode(blockHash, blockNumber, pendingChange, isDescendantOf)
+		if err != nil {
+			return false, err
+		}
+
+		if imported {
+			return true, nil
+		}
+	}
+
+	isDescendant, err := isDescendantOf(announcingHash, blockHash)
+	if err != nil {
+		return false, fmt.Errorf("cannot check ancestry: %w", err)
+	}
+
+	if !isDescendant {
+		return false, nil
+	}
+
+	childrenNode := &pendingChangeNode{change: pendingChange}
+	c.nodes = append(c.nodes, childrenNode)
+	return true, nil
+}
+
+// changeTree keeps track of the changes per fork allowing
+// n forks in the same structure, this structure is intended
+// to be an acyclic directed graph where the change nodes are
+// placed by descendency order and number, you can ensure an
+// node ancestry using the `isDescendantOfFunc`
+type changeTree []*pendingChangeNode
+
+func (ct changeTree) Len() int { return len(ct) }
+func (ct *changeTree) importChange(pendingChange *pendingChange, isDescendantOf isDescendantOfFunc) error {
+	for _, root := range *ct {
+		imported, err := root.importNode(pendingChange.announcingHeader.Hash(),
+			pendingChange.announcingHeader.Number, pendingChange, isDescendantOf)
+
+		if err != nil {
+			return err
+		}
+
+		if imported {
+			logger.Debugf("changes on header %s (%d) imported successfully",
+				pendingChange.announcingHeader.Hash(), pendingChange.announcingHeader.Number)
+			return nil
+		}
+	}
+
+	pendingChangeNode := &pendingChangeNode{
+		change: pendingChange,
+	}
+
+	*ct = append(*ct, pendingChangeNode)
+	return nil
+}
+
+// lookupChangesWhere returns the first change which satisfy the
+// condition whithout modify the current state of the change tree
+func (ct changeTree) lookupChangeWhere(condition conditionFunc[*pendingChangeNode]) (
```

Comment:
why do we need to use generics in this case? Seems like all calls of `lookupChangeWhere` would only accept type `*pendingChangeNode`, so why not simplify and accept a function with signature `func (n *pendingChangeNode) (bool, err)`

Reply:
Not really, to lookup in the `orderedPendingChanges` the `lookupChangeWhere` function should iterate over `*pendingChange` and for the `changeTree` the `lookupChangeWhere` function should iterate over `*pendingChangeNode`.

I decided to use the generic function type `type conditionFunc[T any] func(T) (bool, error)` as it is the same as creating the parameter for each lookup function, like: `func(*pendingChangeNode) (bool, error)` and `func(*pendingChange) (bool, error)`